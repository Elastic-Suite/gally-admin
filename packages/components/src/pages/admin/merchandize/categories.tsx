import React, { useContext, useEffect, useMemo, useRef, useState } from 'react'
import { useTranslation } from 'next-i18next'
import { useRouter } from 'next/router'
import { closeSnackbar, enqueueSnackbar } from 'notistack'
import {
  ICategories,
  ICategory,
  ICategoryConfiguration,
  IParsedCategoryConfiguration,
  IRuleCombination,
  LoadStatus,
  isError,
  isRuleValid,
  isVirtualCategoryEnabled,
  parseCatConf,
  savePositions,
  serializeCatConf,
} from '@elastic-suite/gally-admin-shared'

import { breadcrumbContext, catalogContext } from '../../../contexts'
import { withAuth, withOptions } from '../../../hocs'
import {
  useApiFetch,
  useApiGraphql,
  useFetchApi,
  useResource,
  useResourceOperations,
  useRuleEngineGraphqlFilters,
  useRuleOperators,
} from '../../../hooks'
import { findCategory } from '../../../services'
import { selectBundles, useAppSelector } from '../../../store'

import Placeholder from '../../../components/atoms/Placeholder/Placeholder'
import TitleBlock from '../../../components/molecules/layout/TitleBlock/TitleBlock'
import TwoColsLayout from '../../../components/molecules/layout/twoColsLayout/TwoColsLayout'
import CatalogSwitcher from '../../../components/stateful/CatalogSwitcher/CatalogSwitcher'
import CategoryTree from '../../../components/stateful/CategoryTree/CategoryTree'
import ProductsContainer from '../../../components/stateful/ProductsContainer/ProductsContainer'
import RulesManager from '../../../components/stateful/RulesManager/RulesManager'

const pagesSlug = ['merchandize', 'categories']

function AdminMerchandizeCategories(): JSX.Element {
  const router = useRouter()
  const fetchApi = useApiFetch()
  const { t } = useTranslation('categories')
  const [isLoading, setIsLoading] = useState(false)
  const bundles = useAppSelector(selectBundles)
  const { catalogId, localizedCatalogId } = useContext(catalogContext)

  // Breadcrumb
  const [, setBreadcrumb] = useContext(breadcrumbContext)
  useEffect(() => {
    setBreadcrumb(pagesSlug)
  }, [router.query, setBreadcrumb])

  // Rule engine operators
  const ruleOperators = useRuleOperators()

  // Categories
  const [selectedCategoryItem, setSelectedCategoryItem] = useState<ICategory>()
  const filters = useMemo(() => {
    const filters: { catalogId?: number; localizedCatalogId?: number } = {}
    if (catalogId && localizedCatalogId) {
      if (catalogId !== -1) {
        filters.catalogId = catalogId
      }
      if (localizedCatalogId !== -1) {
        filters.localizedCatalogId = localizedCatalogId
      }
    }

    return filters
  }, [catalogId, localizedCatalogId])
  const [categories] = useFetchApi<ICategories>('categoryTree', filters)
  useEffect(() => {
    if (categories.status !== LoadStatus.SUCCEEDED) {
      return
    }
    setSelectedCategoryItem((prevState) => {
      if (!categories.data.categories[0]) {
        return undefined
      }
      if (prevState === undefined) {
        return categories.data.categories[0]
      }
      const cat = findCategory(prevState, categories.data.categories)
      return cat
    })
  }, [categories])

  // Category configuration
  const catConfResource = useResource('CategoryConfiguration')
  const [catConf, setCatConf] = useState<IParsedCategoryConfiguration>()
  const prevCatConf = useRef<IParsedCategoryConfiguration>()
  const { update, create } =
    useResourceOperations<ICategoryConfiguration>(catConfResource)
  useEffect(() => {
    if (
      selectedCategoryItem?.id &&
      categories.status === LoadStatus.SUCCEEDED
    ) {
      if (
        categories?.data?.categories.find(
          (item) => item.id === selectedCategoryItem?.id
        )
      ) {
        fetchApi<ICategoryConfiguration>(
          `${catConfResource.url}/category/${selectedCategoryItem.id}`,
          filters
        ).then((catConf) => {
          if (!isError(catConf)) {
            const parsedCatConf = parseCatConf(catConf)
            prevCatConf.current = parsedCatConf
            setCatConf(parsedCatConf)
          }
        })
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [fetchApi, selectedCategoryItem?.id, catConfResource.url, categories])

  const isValid = !catConf?.isVirtual || isRuleValid(catConf?.virtualRule)

  // Rule engine graphql filters
  const [productGraphqlFilters, setRuleEngineGraphqlFilters] =
    useRuleEngineGraphqlFilters(catConf, ruleOperators, selectedCategoryItem)

  // Product positions
  const prevProductPositions = useRef<string>('')

  function handleUpdateCat(name: string, val: boolean | string): void {
    setCatConf((catConf) => ({ ...catConf, [name]: val }))
  }

  function handleUpdateRule(rule: IRuleCombination): void {
    setCatConf((catConf) => ({ ...catConf, virtualRule: rule }))
  }

  function handleSelectCategory(category: ICategory): void {
    setSelectedCategoryItem(category)
    setCatConf(null)
    setRuleEngineGraphqlFilters(null)
  }

  const graphqlApi = useApiGraphql()
  async function saveProductPositions(positions: string): Promise<void> {
    const variables = {
      categoryId: selectedCategoryItem?.id,
      savePositionsCategory: positions,
    }
    const json = await graphqlApi(savePositions, variables)
    if (!isError(json)) {
      prevProductPositions.current = positions
    } else {
      throw new Error(json.error.message)
    }
  }

  async function saveCategoryConfiguration(): Promise<void> {
    const serializedCatConf = serializeCatConf(catConf, ruleOperators)
    if (!catConf.id) {
      delete serializedCatConf['@id']
      const newCatConf = await create(serializedCatConf)
      if (!isError(newCatConf)) {
        const parsedCatConf = parseCatConf(newCatConf)
        prevCatConf.current = parsedCatConf
        setCatConf(parsedCatConf)
      } else {
        throw new Error(newCatConf.error.message)
      }
    } else {
      const newCatConf = await update(serializedCatConf.id, serializedCatConf)
      if (!isError(newCatConf)) {
        const parsedCatConf = parseCatConf(newCatConf)
        prevCatConf.current = parsedCatConf
        setCatConf(parsedCatConf)
      } else {
        throw new Error(newCatConf.error.message)
      }
    }
  }

  async function onSave(result: string): Promise<void> {
    setIsLoading(true)
    const promiseGraphQlSetPositions = saveProductPositions(result)
    const promiseApiRestVal = saveCategoryConfiguration()
    try {
      await Promise.all([promiseGraphQlSetPositions, promiseApiRestVal])
      enqueueSnackbar(t('alert'), {
        onShut: closeSnackbar,
        variant: 'success',
      })
    } catch (error) {
      enqueueSnackbar(t('alert.error'), {
        onShut: closeSnackbar,
        variant: 'error',
      })
    } finally {
      setIsLoading(false)
    }
  }

  return (
    <>
      <TwoColsLayout
        left={[
          <TitleBlock key="categories" title={t(pagesSlug.at(-1))}>
            <>
              <CatalogSwitcher />
              <CategoryTree
                categories={categories.data}
                selectedItem={selectedCategoryItem}
                onSelect={handleSelectCategory}
              />
            </>
          </TitleBlock>,
          isVirtualCategoryEnabled(bundles) && ruleOperators && (
            <TitleBlock
              key="virtualRule"
              title={catConf?.isVirtual ? t('virtualRule.title') : ''}
            >
              <RulesManager
                active={catConf?.isVirtual}
                onChange={handleUpdateRule}
                rule={catConf?.virtualRule}
                ruleOperators={ruleOperators}
              />
            </TitleBlock>
          ),
        ]}
      >
        {Boolean(selectedCategoryItem?.id) && (
          <Placeholder placeholder={t('placeholder')}>
            <ProductsContainer
              catConf={catConf}
              category={selectedCategoryItem}
              isValid={isValid}
              onChange={handleUpdateCat}
              onSave={onSave}
              prevCatConf={prevCatConf}
              prevProductPositions={prevProductPositions}
              productGraphqlFilters={productGraphqlFilters}
              isLoading={isLoading}
            />
          </Placeholder>
        )}
      </TwoColsLayout>
    </>
  )
}

export default withAuth(withOptions(AdminMerchandizeCategories))
