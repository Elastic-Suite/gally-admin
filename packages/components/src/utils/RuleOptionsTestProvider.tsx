import React, { ReactNode } from 'react'
import {
  IRuleEngineOperators,
  RuleAttributeType,
  getIri,
} from '@elastic-suite/gally-admin-shared'

import ruleEngineOperator from '../../public/mocks/rule_engine_operators.json'
import sourceFields from '../../public/mocks/source_fields.json'
import sourceFieldLabels from '../../public/mocks/source_field_labels.json'

import RuleOptionsProvider from '../components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactNode
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(getIri('source_fields', field.id))
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider
      catalogId={-1}
      localizedCatalogId={-1}
      fields={fields}
      ruleOperators={ruleEngineOperator as IRuleEngineOperators}
    >
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
