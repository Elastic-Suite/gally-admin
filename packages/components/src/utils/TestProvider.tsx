import React, { ReactNode } from 'react'

import catalogs from '../../public/mocks/catalog.json'
import metadata from '../../public/mocks/metadata.json'

import { catalogContext, metadataContext } from '../contexts'
import { useCatalogs } from '../hooks'

import OptionsProvider from '../components/stateful-providers/OptionsProvider/OptionsProvider'

interface IProps {
  children: ReactNode
}

function TestProvider(props: IProps): JSX.Element {
  const { children } = props
  const contextValueCatalog = useCatalogs(catalogs['hydra:member'])
  const contextValueMetadata = metadata['hydra:member']

  return (
    <OptionsProvider>
      <catalogContext.Provider value={contextValueCatalog}>
        <metadataContext.Provider value={contextValueMetadata}>
          {children}
        </metadataContext.Provider>
      </catalogContext.Provider>
    </OptionsProvider>
  )
}

export default TestProvider
