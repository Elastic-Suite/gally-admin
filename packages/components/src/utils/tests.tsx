import React, { ReactNode } from 'react'
import { RenderOptions, render, renderHook } from '@testing-library/react'
import { PreloadedState } from '@reduxjs/toolkit'
import { Bundle, api } from '@elastic-suite/gally-admin-shared'

import { AppStore, RootState, setupStore } from '../store'

import AppProvider from '../components/stateful-providers/AppProvider/AppProvider'

import TestProvider from './TestProvider'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

const defaultState = {
  data: {
    api,
    bundles: [Bundle.VIRTUAL_CATEGORY],
    configurations: {
      'base_url/media': 'https://localhost/media/catalog/product/',
    },
    metadata: [
      {
        '@id': '/metadata/3',
        '@type': 'Metadata',
        id: 3,
        entity: 'product',
        sourceFields: ['/source_fields/217', '/source_fields/253'],
      },
      {
        '@id': '/metadata/4',
        '@type': 'Metadata',
        id: 4,
        entity: 'category',
        sourceFields: ['/source_fields/254', '/source_fields/286'],
      },
    ],
  },
  user: {
    requestedPath: '',
  },
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = defaultState,
    store,
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  store = store ?? setupStore(preloadedState)
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <AppProvider store={store}>
        <TestProvider>{children}</TestProvider>
      </AppProvider>
    )
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
