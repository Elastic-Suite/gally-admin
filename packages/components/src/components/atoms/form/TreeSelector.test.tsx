import React from 'react'

import { renderWithProviders } from '../../../utils/tests'

import TreeSelector from './TreeSelector'
import { categories } from '../../../../public/mocks/categories.json'

import { screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

describe('TreeSelectorError', () => {
  it('sould match snapshot', () => {
    const { container } = renderWithProviders(
      <TreeSelector
        label="Label"
        data={categories}
        value={{
          id: 'two',
          name: 'Catégorie Deux',
          level: 1,
          path: 'two',
          isVirtual: false,
        }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('sould match snapshot width multiple value', () => {
    const { container } = renderWithProviders(
      <TreeSelector
        label="Label"
        multiple
        data={categories}
        value={[
          {
            id: 'two',
            name: 'Catégorie Deux',
            level: 1,
            path: 'two',
            isVirtual: false,
          },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should display error with showError prop when the field is required with missing value', () => {
    renderWithProviders(
      <TreeSelector
        label="Label"
        data={categories}
        required
        showError
        multiple
        value={[]}
      />
    )
    expect(screen.getByText('formError.valueMissing')).toBeInTheDocument()
  })

  it('should not display error without showError prop when the field is required with missing value', () => {
    renderWithProviders(
      <TreeSelector
        label="Label"
        multiple
        data={categories}
        value={[]}
        required
      />
    )
    expect(screen.queryByText('formError.valueMissing')).not.toBeInTheDocument()
  })

  it('should not display error with showError prop when the field is required and disabled with missing value', () => {
    renderWithProviders(
      <TreeSelector
        label="Label"
        data={categories}
        value={[]}
        multiple
        required
        showError
        disabled
      />
    )
    expect(screen.queryByText('formError.valueMissing')).not.toBeInTheDocument()
  })

  it('shoud display error with showError prop when the field has additional validation rules', () => {
    renderWithProviders(
      <TreeSelector
        label="Label"
        data={categories}
        value={[]}
        multiple
        required
        showError
        additionalValidator={(): string => 'erreur'}
      />
    )
    expect(screen.getByText('formError.erreur')).toBeInTheDocument()
  })
})
