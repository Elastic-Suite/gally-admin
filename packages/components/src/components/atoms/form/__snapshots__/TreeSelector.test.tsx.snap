// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeSelector sould match snapshot 1`] = `
<div>
  <div
    class="css-1b4s9jn"
  >
    <div
      class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled css-rnc4nv-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r0:"
      >
        Label
      </label>
      <div
        class="InputText__Wrapper css-17fp4jk"
      >
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd hasClearIcon css-4g1i97-MuiInputBase-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiInputBase-inputAdornedEnd css-ngazp5-MuiInputBase-input"
            id=":r0:"
            role="combobox"
            spellcheck="false"
            type="text"
            value="Catégorie Deux"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1uab8qk-MuiInputAdornment-root"
          >
            <button
              aria-label="form.clear"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium AutocompleteClearIndicator css-3rozpy-MuiButtonBase-root-MuiIconButton-root"
              tabindex="-1"
              title="form.clear"
              type="button"
            >
              <ion-icon
                src="/images/ionicons/close.svg"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="form.open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1x878dy-MuiButtonBase-root-MuiIconButton-root"
              tabindex="-1"
              title="form.open"
              type="button"
            >
              <ion-icon
                src="/images/ionicons/chevron-down.svg"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TreeSelector sould match snapshot width multiple value 1`] = `
<div>
  <div
    class="css-1b4s9jn"
  >
    <div
      class="MuiFormControl-root css-hh3vi8-MuiFormControl-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary css-rnc4nv-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r1:"
      >
        Label
      </label>
      <div
        class="InputText__Wrapper css-17fp4jk"
      >
        <div
          class="MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd hasClearIcon css-4g1i97-MuiInputBase-root"
        >
          <div
            class="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-deletable MuiChip-deletableColorDefault MuiChip-filledDefault css-69f6vo-MuiButtonBase-root-MuiChip-root"
            data-tag-index="0"
            role="button"
            style="opacity: 1;"
            tabindex="-1"
          >
            <span
              class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
            >
              Catégorie Deux
            </span>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiChip-deleteIcon MuiChip-deleteIconMedium MuiChip-deleteIconColorDefault MuiChip-deleteIconFilledColorDefault css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-testid="chipCloseButton"
              tabindex="0"
              type="button"
            >
              <ion-icon
                src="/images/ionicons/close.svg"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-ngazp5-MuiInputBase-input"
            id=":r1:"
            role="combobox"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1uab8qk-MuiInputAdornment-root"
          >
            <button
              aria-label="form.clear"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium AutocompleteClearIndicator css-3rozpy-MuiButtonBase-root-MuiIconButton-root"
              tabindex="-1"
              title="form.clear"
              type="button"
            >
              <ion-icon
                src="/images/ionicons/close.svg"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="form.open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1x878dy-MuiButtonBase-root-MuiIconButton-root"
              tabindex="-1"
              title="form.open"
              type="button"
            >
              <ion-icon
                src="/images/ionicons/chevron-down.svg"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
